File: app/src/main/java/com/beautifulquran/ui/home/ContinueInk.kt
package com.beautifulquran.ui.home

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.CubicBezierEasing
import androidx.compose.animation.core.tween
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.snapshotFlow
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.runtime.withFrameNanos
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clipToBounds
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.graphics.BlendMode
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Paint
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.drawIntoCanvas
import androidx.compose.ui.unit.dp
import com.beautifulquran.ui.PageTurnSounds
import com.beautifulquran.ui.theme.LocalNuqtaParams
import com.beautifulquran.ui.theme.NuqtaParams
import com.beautifulquran.ui.theme.drawInkWashCoats
import com.beautifulquran.ui.theme.inkCoatFingers
import com.beautifulquran.ui.theme.inkWashReachToCover
import kotlin.math.hypot
import kotlin.math.roundToInt
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.launch

/**
 * How far the paper stack has turned from the chapter list into the open
 * reader, 0..1 — 0 when no reader is open to turn to. Read it only while
 * drawing or in snapshot flows: it changes every frame of a swipe.
 */
val LocalReaderApproach = staticCompositionLocalOf<() -> Float> { { 0f } }

/**
 * How long the page-turn's current sweep spike lasts, in ms. The continue
 * row's wipe ends with that spike, not with the stem's quiet tail.
 */
val LocalSweepSpikeMs = staticCompositionLocalOf<() -> Int> {
    { PageTurnSounds.FLIPS[1].sweepSpikeEndMs }
}

/**
 * Ink that floods the continue row the moment the chapter list starts to turn
 * into the reader it continues. The wash plays out on its own clock whatever
 * the drag does. It stays wet under the reader. Returning, it stays wet
 * through the lift and wipes with the sweep, ending when that stem's volume
 * spike ends. A turn let go short wipes the same way. The ink stays dense as
 * it withdraws, so every word is plainly on ink or on paper, never half-way.
 */
@Stable
internal class ContinueInk(val params: NuqtaParams) {
    /** How far the wash has spread, 0..1, eased. Read it only while drawing. */
    var spread by mutableFloatStateOf(0f)
        private set

    /** The wash's own clock, 0..1, before easing. */
    private var clock = 0f
    /** How far the wash has wiped back out of the row, 0..1, eased. */
    val dry = Animatable(0f)

    /** Tapped, and the turn it starts has not yet landed. */
    var tapped by mutableStateOf(false)

    val fingers = inkCoatFingers(params)

    /**
     * Runs the wash to full on its own clock. Each frame advances it by at
     * most [ContinueInkMaxFrameMs], so a long frame — the reader being built
     * as the turn starts — pauses the wash instead of skipping its flood.
     */
    suspend fun flood() {
        var last = withFrameNanos { it }
        while (clock < 1f) {
            withFrameNanos { now ->
                val ms = ((now - last) / 1_000_000f).coerceAtMost(ContinueInkMaxFrameMs)
                last = now
                clock = (clock + ms / ContinueInkSpreadMs).coerceAtMost(1f)
                spread = ContinueInkEasing.transform(clock)
            }
        }
    }

    /** Fully inked at once, for a row under the reader. */
    fun fill() {
        clock = 1f
        spread = 1f
    }

    fun clear() {
        clock = 0f
        spread = 0f
    }
}

private enum class ContinueTurn { Resting, Turning, Landed }

@Composable
internal fun rememberContinueInk(): ContinueInk {
    val params = LocalNuqtaParams.current
    val ink = remember(params) { ContinueInk(params) }
    val approach = LocalReaderApproach.current
    val sweepSpikeMs = LocalSweepSpikeMs.current
    LaunchedEffect(ink, approach, sweepSpikeMs) {
        var fromReader = false
        snapshotFlow {
            val p = approach()
            when {
                p >= 1f - ContinueTurnEpsilon -> ContinueTurn.Landed
                p > ContinueTurnEpsilon || ink.tapped -> ContinueTurn.Turning
                else -> ContinueTurn.Resting
            }
        }.distinctUntilChanged().collectLatest { turn ->
            when (turn) {
                ContinueTurn.Turning -> coroutineScope {
                    if (fromReader && !ink.tapped) {
                        // Stay wet through the lift. The wipe starts with the
                        // sweep stem and runs only as long as that stem's
                        // volume spike, so it ends when the whoosh does.
                        snapshotFlow { approach() <= 1f - PageTurnSounds.SWEEP_AT }
                            .distinctUntilChanged()
                            .collectLatest { sweeping ->
                                if (sweeping) {
                                    wipeContinueInk(ink, sweepSpikeMs())
                                } else if (ink.dry.value > 0f) {
                                    ink.dry.animateTo(
                                        0f,
                                        tween(
                                            (sweepSpikeMs() * ink.dry.value).roundToInt(),
                                            easing = ContinueInkRewetEasing,
                                        ),
                                    )
                                }
                            }
                    } else {
                        fromReader = false
                        // The wash plays out on its own clock. Turned again
                        // mid-wipe, the wipe runs back from where it stands, as
                        // fast as it came, while the flood carries on — every
                        // interruption picks up the ink exactly as it is.
                        launch {
                            ink.dry.animateTo(
                                0f,
                                tween(
                                    (sweepSpikeMs() * ink.dry.value).roundToInt(),
                                    easing = ContinueInkRewetEasing,
                                ),
                            )
                        }
                        ink.flood()
                    }
                }
                ContinueTurn.Landed -> {
                    // Under the reader the row stays wet, so turning back
                    // uncovers it inked and it dries with the sweep.
                    ink.tapped = false
                    fromReader = true
                    ink.fill()
                    ink.dry.snapTo(0f)
                }
                ContinueTurn.Resting -> {
                    fromReader = false
                    if (ink.spread > 0f) {
                        wipeContinueInk(ink, sweepSpikeMs())
                        ink.clear()
                        ink.dry.snapTo(0f)
                    }
                }
            }
        }
    }
    return ink
}

private suspend fun wipeContinueInk(ink: ContinueInk, spikeMs: Int) {
    // A wipe interrupted and resumed covers only what is left of it, in
    // its share of the time — ending with the spike, not after it.
    val wipeEasing = if (ink.dry.value > 0f) ContinueInkRewetEasing else ContinueInkWipeEasing
    ink.dry.animateTo(
        1f,
        tween(continueWipeMs(spikeMs, ink.dry.value), easing = wipeEasing),
    )
}

/** Remaining wipe, so a resumed dry still ends with the spike. */
internal fun continueWipeMs(spikeMs: Int, dry: Float): Int =
    (spikeMs * (1f - dry)).roundToInt().coerceAtLeast(0)

/** The wash itself, drawn behind the row, landing at its right (Arabic) end. */
internal fun Modifier.continueInkWash(ink: ContinueInk, color: Color): Modifier =
    clipToBounds().drawBehind { drawContinueInk(ink, color) }

/**
 * For a copy of the row's content in colours that read on the ink: shows it
 * only where the wash lies, and exactly as strongly as the wash is dense.
 */
internal fun Modifier.continueInkMask(ink: ContinueInk): Modifier = drawWithContent {
    if (ink.spread <= 0f) return@drawWithContent
    drawIntoCanvas { canvas ->
        val bounds = Rect(Offset.Zero, size)
        canvas.saveLayer(bounds, Paint())
        drawContent()
        // Its own layer so, composited DstIn, bare paper clears the copy too.
        canvas.saveLayer(bounds, Paint().apply { blendMode = BlendMode.DstIn })
        drawContinueInk(ink, Color.Black)
        canvas.restore()
        canvas.restore()
    }
}

private fun DrawScope.drawContinueInk(ink: ContinueInk, color: Color) {
    val origin = Offset(size.width, size.height / 2f)
    fun wash() = drawInkWashCoats(
        t = ink.spread,
        dry = 0f,
        params = ink.params,
        fingers = ink.fingers,
        ink = color,
        origin = origin,
        full = inkWashReachToCover(hypot(size.width, size.height / 2f)),
    )
    val wipe = ink.dry.value
    if (wipe <= 0f) {
        wash()
        return
    }
    // Wiping back: a soft edge sweeps from the row's far end toward where the
    // ink landed, and only what lies behind it keeps its ink.
    val feather = ContinueInkWipeFeather.toPx()
    val edge = -feather + (size.width + 2f * feather) * wipe
    drawIntoCanvas { canvas ->
        canvas.saveLayer(Rect(Offset.Zero, size), Paint())
        wash()
        drawRect(
            brush = Brush.horizontalGradient(
                0f to Color.Transparent,
                1f to Color.Black,
                startX = edge - feather,
                endX = edge + feather,
            ),
            blendMode = BlendMode.DstIn,
        )
        canvas.restore()
    }
}

/** The wash's own clock, from the turn starting to the row fully inked. */
private const val ContinueInkSpreadMs = 420

/**
 * The wash floods out fast and eases long into the row's far end: a soft
 * start so it never pops, then most of the row in the first third.
 */
private val ContinueInkEasing = CubicBezierEasing(0.3f, 0f, 0.2f, 1f)

private const val ContinueTurnEpsilon = 0.001f

/** The most a single frame may advance the wash's clock. */
private const val ContinueInkMaxFrameMs = 20f

/** The wipe gathers, crosses the words briskly, and settles into the Arabic end. */
private val ContinueInkWipeEasing = CubicBezierEasing(0.4f, 0f, 0.2f, 1f)

/** Half the width of the wipe's soft edge. */
private val ContinueInkWipeFeather = 14.dp

/**
 * For a wipe picking up mid-way, in either direction: it is already moving,
 * so it starts at speed and settles, never pausing to gather.
 */
private val ContinueInkRewetEasing = CubicBezierEasing(0f, 0f, 0.2f, 1f)

File: web/src/ui/home/continueInk.ts
import { cubicBezierEase } from '../theme/Fade'

/** Page-turn stems — Android `PageTurnSounds`. Wipe is scrubbed to these. */
export const SWEEP_AT = 0.42
export const DROP_AT = 0.88

/** A full sheet turn. Android `STACK_PAGE_DURATION_MS`. */
export const SHEET_TURN_MS = 460

const CONTINUE_TURN_EPSILON = 0.001
const CONTINUE_INK_SPREAD_MS = 420
const CONTINUE_INK_MAX_FRAME_MS = 20
/** Leftover wipes cover the sweep-to-drop window of a 460ms turn. */
export const CONTINUE_INK_DRY_MS = 212

const FLOOD_EASE = [0.3, 0, 0.2, 1] as const
const WIPE_EASE = [0.4, 0, 0.2, 1] as const
const REWET_EASE = [0, 0, 0.2, 1] as const

export type ContinueTurn = 'resting' | 'turning' | 'landed'

export function continueInkToContinueChapter(openSurahId: number, continueSurahId: number): boolean {
  return continueSurahId !== 0 && openSurahId === continueSurahId
}

/** Continue was chosen: a tap, or a swipe into that chapter. */
export function continueInkShouldFlood(
  tapped: boolean,
  fromReader: boolean,
  toContinueChapter: boolean,
): boolean {
  return !fromReader && (tapped || toContinueChapter)
}

/** Returning from a Continue-owned reader: wipe with the page-turn drop. */
export function continueInkShouldWipe(tapped: boolean, fromReader: boolean): boolean {
  return fromReader && !tapped
}

/** Landed on a Continue-owned reader: keep the row wet underneath. */
export function continueInkShouldFill(
  tapped: boolean,
  spread: number,
  toContinueChapter: boolean,
): boolean {
  return tapped || spread > 0 || toContinueChapter
}

/**
 * How far the continue row has wiped for a return swipe of [turnProgress]
 * (0 at the reader, 1 at home). 0 through the lift, 1 the frame the drop begins.
 */
export function continueWipeProgress(turnProgress: number): number {
  return clamp((turnProgress - SWEEP_AT) / (DROP_AT - SWEEP_AT))
}

/** Remaining leftover wipe, as a share of the sweep-to-drop window. */
export function continueWipeMs(dry: number): number {
  return Math.max(0, Math.round(CONTINUE_INK_DRY_MS * (1 - dry)))
}

function clamp(x: number, lo = 0, hi = 1): number {
  return Math.min(hi, Math.max(lo, x))
}

function ease(t: number, curve: readonly [number, number, number, number]): number {
  return cubicBezierEase(clamp(t), curve[0], curve[1], curve[2], curve[3])
}

type DryAnim = {
  from: number
  to: number
  elapsed: number
  dur: number
  curve: readonly [number, number, number, number]
  onEnd?: () => void
}

/**
 * The continue row's wash, without React. Floods on its own 420ms clock,
 * stays wet under a Continue-owned reader, and wipes back scrubbed to the
 * page-turn drop. Opening some other chapter leaves the row dry.
 */
export class ContinueInkMachine {
  spread = 0
  dry = 0
  tapped = false

  private clock = 0
  private fromReader = false
  private turn: ContinueTurn = 'resting'
  private flooding = false
  private scrubbing = false
  private dryAnim: DryAnim | null = null

  /** True while a flood, wipe, or scrub still has frames to paint. */
  busy(): boolean {
    return this.flooding || this.scrubbing || this.dryAnim != null || this.tapped
  }

  step(dtMs: number, approach: number, openSurahId: number, continueSurahId: number): void {
    const next = this.classify(approach)
    if (next !== this.turn) this.enter(next, approach, openSurahId, continueSurahId)
    if (this.scrubbing) {
      this.dry = continueWipeProgress(1 - approach)
    } else {
      this.advanceDry(dtMs)
    }
    if (this.flooding) this.advanceFlood(dtMs)
  }

  private classify(approach: number): ContinueTurn {
    if (approach >= 1 - CONTINUE_TURN_EPSILON) return 'landed'
    if (approach > CONTINUE_TURN_EPSILON || this.tapped) return 'turning'
    return 'resting'
  }

  private enter(
    turn: ContinueTurn,
    approach: number,
    openSurahId: number,
    continueSurahId: number,
  ): void {
    this.turn = turn
    this.dryAnim = null
    this.scrubbing = false
    this.flooding = false
    const toContinue = continueInkToContinueChapter(openSurahId, continueSurahId)
    if (turn === 'turning') {
      if (continueInkShouldWipe(this.tapped, this.fromReader)) {
        this.scrubbing = true
        this.dry = continueWipeProgress(1 - approach)
      } else if (continueInkShouldFlood(this.tapped, this.fromReader, toContinue)) {
        this.fromReader = false
        this.flooding = true
        this.startDry(0, CONTINUE_INK_DRY_MS * this.dry, REWET_EASE)
      }
      return
    }
    if (turn === 'landed') {
      const owned = continueInkShouldFill(this.tapped, this.spread, toContinue)
      this.tapped = false
      if (owned) {
        this.fromReader = true
        this.clock = 1
        this.spread = 1
        this.dry = 0
      } else {
        this.fromReader = false
      }
      return
    }
    this.fromReader = false
    if (this.spread <= 0) return
    if (this.dry < 1) {
      const curve = this.dry > 0 ? REWET_EASE : WIPE_EASE
      this.startDry(1, continueWipeMs(this.dry), curve, () => {
        this.clock = 0
        this.spread = 0
        this.dry = 0
      })
    } else {
      this.clock = 0
      this.spread = 0
      this.dry = 0
    }
  }

  private startDry(
    to: number,
    dur: number,
    curve: readonly [number, number, number, number],
    onEnd?: () => void,
  ): void {
    if (dur <= 0) {
      this.dry = to
      onEnd?.()
      return
    }
    this.dryAnim = { from: this.dry, to, elapsed: 0, dur, curve, onEnd }
  }

  private advanceDry(dtMs: number): void {
    const anim = this.dryAnim
    if (!anim) return
    anim.elapsed += dtMs
    const u = clamp(anim.elapsed / anim.dur)
    this.dry = anim.from + (anim.to - anim.from) * ease(u, anim.curve)
    if (u >= 1) {
      this.dry = anim.to
      this.dryAnim = null
      anim.onEnd?.()
    }
  }

  private advanceFlood(dtMs: number): void {
    if (this.clock >= 1) {
      this.spread = 1
      this.flooding = false
      return
    }
    const ms = Math.min(Math.max(dtMs, 0), CONTINUE_INK_MAX_FRAME_MS)
    this.clock = Math.min(1, this.clock + ms / CONTINUE_INK_SPREAD_MS)
    this.spread = ease(this.clock, FLOOD_EASE)
    if (this.clock >= 1) this.flooding = false
  }
}
